aten/src/ATen/native/NativeFunctions.cpp

```diff
@@ -584,5 +586,77 @@ Tensor RoiPooling2d_backward_cpu(
   throw std::runtime_error("not implemented");
 }
 
+
+
+// TODO Replace this with more accurate digamma().
+template <typename Scalar>
+static inline Scalar digamma_one(Scalar x) {
+  const Scalar eps = x * 1e-2;
+  return (std::lgamma(x + eps) - std::lgamma(x - eps)) / (eps + eps);
+}
+
+/** Computes the reparameterized gradient -(d/dalpha cdf(x;alpha)) / pdf(x;alpha)
+    for random number x drawn from a standard Gamma distribution Gamma(alpha).
+*/
+template <typename Scalar>
+static inline Scalar standard_gamma_grad_one(Scalar x, Scalar alpha) {
+  // Use an asymptotic approximation for small x.
+  if (x < 0.2f) {
+    const Scalar a0 = 1 / alpha;
+    const Scalar a1 = 1 / (alpha + 1);
+    const Scalar a2 = 1 / (alpha + 2);
+    const Scalar pow_x_alpha = std::pow(x, alpha);
+    const Scalar gamma_pdf = std::pow(x, alpha - 1) * std::exp(-x);
+    const Scalar gamma_cdf = pow_x_alpha * (a0 - x*a1 + 0.5f*x*x*a2);
+    const Scalar gamma_cdf_alpha = (std::log(x) - digamma_one(alpha)) * gamma_cdf
+        - pow_x_alpha * (a0*a0 - x*a1*a1 + 0.5f*x*x*a2*a2);
+    const Scalar result = -gamma_cdf_alpha / gamma_pdf;
+    return std::isnan(result) ? 0 : result;
+  }
+
+  // Use an asymptotic approximation for large alpha.
+  if (alpha > 50.0f) {
+    return std::sqrt(x / alpha);
+  }
+
+  // Use a bivariate rational approximation to the reparameterized gradient.
+  const Scalar u = std::log(x / alpha);
+  const Scalar v = std::log(alpha);
+  static const Scalar coef_uv[3][8] = {
+    {0.16028008, -0.088064309, 0.019630876, -0.0016920282,
+     1.0, 0.36659853, 0.10843863, 0.0066895454},
+    {0.521894, 0.16095838, 0.06237597, 0.0023884253,
+     0.083457714, 0.0073297628, -0.0059299053, -0.00093720389},
+    {-0.0031143957, -0.012143877, -0.0057656484, -0.00064847254,
+     0.0087262576, -0.00022820524, 1.8871047e-05, 9.6307964e-06},
+  };
+  Scalar coef_v[8];
+  for (int i = 0; i < 8; ++ i) {
+    coef_v[i] = coef_uv[0][i] + u * (coef_uv[1][i] + u * coef_uv[2][i]);
+  }
+  const Scalar p = coef_v[0] + v * (coef_v[1] + v * (coef_v[2] + v * coef_v[3]));
+  const Scalar q = coef_v[4] + v * (coef_v[5] + v * (coef_v[6] + v * coef_v[7]));
+  return std::exp(p / q);
+}
+
+template <typename Scalar>
+struct StandardGammaGradOp {
+  void operator()(Scalar& ret_val, const Scalar& self_val, const Scalar &alpha_val, bool& early_exit)
+  {
+    ret_val = standard_gamma_grad_one(self_val, alpha_val);
+  }
+
+  static void apply(Tensor& ret, const Tensor& self, const Tensor& alpha) {
+    StandardGammaGradOp<Scalar> op;
+    CPU_tensor_apply3<Scalar, StandardGammaGradOp<Scalar>>(ret, self, alpha, op);
+  }
+};
+
+Tensor _standard_gamma_grad(const Tensor& self, const Tensor& alpha) {
+  Tensor ret = self.type().tensor(self.sizes());
+  dispatch_cpu_floating_types<StandardGammaGradOp>(self.type(), "_standard_gamma_grad", ret, self, alpha);
```

Comment:
Ok, I think this is getting more complicated than it needs to be because dispatch was already very complicated.
I think we can make all of this simpler by basically simplifying the dispatch macros to only care about switching over scalar type. The old dispatch can be deleted since it is not used as far as I know.

For instance, here it would be much simpler if.
(1) We used a function EnsureBackend({self, alpha}, at::kCPU) rather than rely on folding it into one function.
(2) We use templated functions rather the classes. The reason for the functions was partial specialization. If you need partial specialization, then just write your own switch statement.
(3) We get rid of the generated dispatch code, this was needed because of the cross product of (backend x scalar_type) that made this exceedingly verbose. If we just dispatch on scalar type, each option can be written directly in Dispatch.h.

Reply:
On (1), I think we should do that on check in CPU_apply (we already do!); it needs to be there (it won't work generally if you call it with a GPU tensors) and it makes calling it easier 
because everyone doesn't need to implement their own checking.  I agree with getting rid of the backend check in the Dispatch function itself.

Reply:
on (3), the code already only generates ScalarType-dispatch only, so yes, we should just move this to a file.

Reply:
On (2) I'm not sure I understand; if we want the op to be able to be a lambda, isn't this basically as simple as we can get?